### DISCUSSION TITLE
When using torchmetrics does pytorch-lightning averages the previous metrics

### DISCUSSION THREAD

**Discussion by yashrathi-git** [Lightning-AI/pytorch-lightning#12187] [General]
By default lightning logs every n steps. So in the nth step does the metrics displayed the average of previous n minibatches or just the accuracy of this specific minibatch?

**Comment by rohitgr7** [comment]
accuracy of the specific batch. the metric is calculated at every step. `log_every_n_steps` is just a barrier that allows that value at that step to pass through the loggers.

**Reply by yashrathi-git** [reply]
why do the logger not avg the prev n metrics, that here seems most sensible thing to do? What is a good way to implement this?

**Reply by rohitgr7** [reply]
you can manually compute it at every steps using some state variables and log it. to create prev n-step logic, you can reset the state after certain steps.

**Reply by yashrathi-git** [reply]
Alright thanks

**Comment by gisilvs** [comment]
Shouldn't this be a feature of Lightning? When I train using iterations and not epochs, I am interested in the average loss value over the last n batches since the last logs, not on the epoch average or current step value.

**Reply by adosar** [reply]
According to the [docs](https://lightning.ai/docs/pytorch/stable/extensions/logging.html#automatic-logging):
* `reduce_fx`: Reduction function over step values for end of epoch. Uses [torch.mean()](https://pytorch.org/docs/stable/generated/torch.mean.html#torch.mean) by default and is not applied when a [torchmetrics.Metric](https://lightning.ai/docs/torchmetrics/stable/references/metric.html#torchmetrics.Metric) is logged.

Does this mean that `torchmetrics.Metric` is not averaged during the end of the epoch?
